### ISSUE TITLE
Jackson dependency brought in by OpenApi

### ISSUE THREAD

**Issue by tomas-langer** [oracle/helidon#1109]
The Small Rye OpenAPI brings in dependency on Jackson:
```
[INFO] |  +- com.fasterxml.jackson.core:jackson-core:jar:2.9.8:compile
[INFO] |  +- com.fasterxml.jackson.core:jackson-databind:jar:2.9.8:compile
[INFO] |  |  \- com.fasterxml.jackson.core:jackson-annotations:jar:2.9.0:compile
[INFO] |  +- com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:jar:2.9.8:compile
```

This is not a desired dependency for Helidon and we should check if this can be replaced with Yasson (which is our preferred JSON-B implementation). If not, we should rebuild the OpenAPI with JSON-P or Yasson dependencies only.

**Comment by tomas-langer** [comment]
BTW this is 523 classes that are added to dependency graph (can be seen in `native-image`)

**Comment by tjquinno** [comment]
We do leverage quite a bit of the smallrye OpenAPI implementation.

It's possible we can continue to use most of that code but provide more of our own logic that will process a static YAML/JSON file. That is the part of OpenAPI that we use that pulls in Jackson.

I'm looking into this now.

**Comment by tjquinno** [comment]
Progress update:

I have the YAML processing using SnakeYAML working, subject to further testing. 

In theory, we can also use SnakeYAML to process JSON, but currently I am running into a problem with an example `petstore.json` file that (reasonably) uses JSON `$ref` to avoid duplicating definitions in the file. SnakeYAML does seemed to work with the JSON file OK up to that point but does not seem to handle `$ref` correctly. I have opened this issue https://bitbucket.org/asomov/snakeyaml/issues/464/support-for-json-ref to see if SnakeYAML does support `$ref` and I am just missing something. I've become pretty familiar with the SnakeYAML source code and I have not seen any logic there that would deal with references, but we'll see what the SnakeYAML folks say.

Plan B could be using JSON-B to handle JSON for our OpenAPI support. That will mean a new set of customizations for JSON-B, parallel to the ones we need for SnakeYAML on the YAML side.

**Comment by tjquinno** [comment]
SnakeYAML indeed does not handle `$ref`s automatically. BTW, Jackson does not handle `$ref`s automatically either -- it's a characteristic of the OpenAPI spec, not of YAML itself -- so I plan to open a separate issue for that (since this is a hole in our current OpenAPI support as a result).

To accomplish that I am looking at ways of extending the SnakeYAML `Constructor` class (which maps the YAML nodes and values to POJOs) to handle the `$ref` features of OpenAPI. 

The outstanding work for this issue is to add some automated checking. We have to provide quite a few customizations via SnakeYAML API calls so SnakeYAML properly processes OpenAPI documents, for every MP OpenAPI interface and implementing class, and in some cases further customizations as well. I want to write some build-time checking as tests that, as much as possible, will make sure that our code covers all the customizations we need. We probably cannot _generate_ that customization code but we can certainly check it for completeness.

**Comment by tjquinno** [comment]
Rather than do automated checking (as discussed in the earlier comment), I have taken the approach of writing a maven plug-in to generate the SnakeYAML helper code, rather than have us write that helper code by hand and have automated checking. 

The plug-in is almost done -- should be ready for review shortly -- and then our OpenAPI module can begin using the plug-in. At that point we should finally be able to cut the cord to Jackson as pulled through by the SmallRye OpenAPI implementation which we layer upon.

**Comment by tjquinno** [comment]
I have merged the plug-in code into the build-tools project. 

The next step is to add use of the plug-in and make a few other changes in the Helidon OpenAPI component itself.

**Comment by tjquinno** [comment]
Even when we bypass the SmallRye YAML parsing itself by doing our own using SnakeYAML, the SmallRye OpenAPI annotation processing still delegated to Jackson directly when handling OpenAPI extensions.

I opened this issue https://github.com/smallrye/smallrye-open-api/issues/245 which gave rise to this PR https://github.com/smallrye/smallrye-open-api/pull/250. It was merged recently, and I have locally built the updated SmallRye code as a snapshot and am using that to progress until there is a SmallRye OpenAPI release that includes the fix. I have asked if there is an ETA for the next release and will update this when I hear back.

**Comment by tjquinno** [comment]
I have created a Helidon-specific `AnnotationScannerExtension` to which the new SmallRye code delegates. The revised Helidon OpenAPI pom excludes Jackson from the SmallRye dependency, and our module builds successfully. 

The remaining step is to change our module to use the generated code from our new SnakeYAML code generator plug-in instead of the hand-coded `Parser` class I started with. (I switched gears to prove out the SmallRye changes to make sure their enhancement meets our needs; it does.)

**Comment by tjquinno** [comment]
I have discovered a few changes that I needed to make in the plug-in, and I have those changes in my fork. I am close to finishing the changes in the Helidon OpenAPI module to use the plug-in but there is a little bit more to do there. Ideally I will hold off on merging the plug-in changes (and requiring another build-tools release) until I have the Helidon OpenAPI changes done and working properly to make sure the plug-in is doing what we need.

**Comment by tjquinno** [comment]
Good progress recently. 

I've discovered two bugs in other code: the MP OpenAPI TCK and SnakeYAML. Here are the issues:

https://github.com/eclipse/microprofile-open-api/issues/412
https://bitbucket.org/asomov/snakeyaml/issues/473/incorrect-scalar-style-chosen-for-float

I've been able to work around these by changing our code.

**Comment by tjquinno** [comment]
Fixed by #1457 (Helidon 1.x) and #1458 (master)
